tests/gen/scenarios/01_deep_delta_generate.run.sh

```diff
@@ -0,0 +1,85 @@
+#!/bin/bash
+# Copyright 2023 Northern.tech AS
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+set -x
+set -e
+
+temp_dir=$(mktemp -d)
+
+function cleanup() {
+    set +e
+    rm -Rf "${temp_dir}"
+}
+
+trap cleanup SIGTERM SIGHUP SIGINT SIGQUIT SIGINT EXIT
+
+artifact_file="${temp_dir}/a0.mender"
+artifact_name=$(basename "$temp_dir")
+"${GENERATOR:-./gen/app-gen}" \
+    --image docker.io/library/debian:10,docker.io/library/debian:latest \
+    --image docker.io/library/postgres:15.0,docker.io/library/postgres:15.1 \
+    --artifact-name "${artifact_name}" \
+    --device-type FP.VT-04 \
+    --output-path "${artifact_file}" \
+    --manifests-dir "${DATA}"/manifests \
+    --orchestrator docker-compose \
+    --application-name myapp4 \
+    --platform linux/arm/v7 \
+    --deep-delta
+
+# lets clean everything.
+docker stop $(docker ps -qa) || true
+docker container rm $(docker container ls -aq) || true
+docker image rmi $(docker image ls -q) || true
+docker system prune -a -f || true
+docker volume rm $(docker volume ls -q) || true
+
+# now we will unpack the artifact
+artifact_dir="${temp_dir}/${artifact_name}"
+mkdir -p "${artifact_dir}"
+tar xvf "${artifact_file}" -C "${artifact_dir}"
+tar xvzf "${artifact_dir}"/data/0000.tar.gz -C "${artifact_dir}"
+tar xvzf "${artifact_dir}"/images.tar.gz -C "${artifact_dir}"
+
+# lets find the postgres:15.1 image
+expected_image_tag=postgres:15.1
+image_file=""
+while read -r; do
+    url=$(cat "$REPLY")
+    if [[ "$url" == "docker.io/library/${expected_image_tag}" ]]; then
+        image="$(dirname "${REPLY}")/image.img"
+        break
+    fi
+done < <(find "${artifact_dir}"/images -name 'url-new.txt')
+[[ "$image" == "" ]] && {
+    echo "assertion failed: cant find image."
+    exit 1
+}
+
+# we will now call LOAD operation from the docker-compose orchestrator submodule
+# we assume that the current image is present, so we will pull it
+docker pull docker.io/library/postgres:15.0 --platform linux/arm/v7
+# and call the submodule, the way the main update module calls
+mkdir -p "${temp_dir}/temp"
+MODULE_TMPDIR="${temp_dir}/temp" OPTIONS="deep_delta" ${DOCKER_COMPOSE_SUBMODULE:-./src/app-modules/docker-compose} \
+    LOAD \
+    myapp \
+    docker.io/library/${expected_image_tag} \
+    "${image}" \
+    docker.io/library/postgres:15.0
+
+diff \
```

Comment:
I don't fully get it. It shows the images are the same. But where can I see that it is actually saving space?

Reply:
I added explicit log and a check that the sizes differ:
```
+ ls -sh /tmp/tmp.gBgMfD/a0.mender /tmp/tmp.gBgMfD/a0-non-deep.mender
 61260 /tmp/tmp.gBgMfD/a0-non-deep.mender
 32972 /tmp/tmp.gBgMfD/a0.mender
++ stat -c %s /tmp/tmp.gBgMfD/a0.mender
+ deep_size=33761792
++ stat -c %s /tmp/tmp.gBgMfD/a0-non-deep.mender
+ regular_size=62729216
+ echo 'delta artifact size: 62729216b'
delta artifact size: 62729216b
+ echo 'deep delta artifact size: 33761792b'
deep delta artifact size: 33761792b
+ [[ 33761792 -lt 62729216 ]]
++ awk -v b=33761792 -v a=62729216 'BEGIN{printf("%0.2lf%%",(a-b)*100.0/((a+b)*0.5));}'
+ p=60.04%
+ echo 'relative percentage gain in deep size: 28967424 (60.04%)'
relative percentage gain in deep size: 28967424 (60.04%)
```

Reply:
Q: What exactly is going on here? Why average the size of both deltas? 

I'm not fully getting what the relative size is measured against. Is it the original Artifact, or is it the original `delta`?

`(a-b)*100.0/((a+b)*0.5))`

Reply:
okey, I changed it to `(a-b)*100.0/a`, so now we are just measuring how much we gained with deep deltas as compared to non-deep. perhaps measuring against average was too much here.